File: binding/src/libqfontcombobox.cpp
#include <QComboBox>
#include <QEvent>
#include <QFont>
#include <QFontComboBox>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QWidget>
#include <qfontcombobox.h>
#include "libqfontcombobox.h"
#include "libqfontcombobox.hxx"

QFontComboBox* QFontComboBox_new(QWidget* parent) {
	 return new VirtualQFontComboBox(parent);
}

QFontComboBox* QFontComboBox_new2() {
	 return new VirtualQFontComboBox();
}

libqt_string QFontComboBox_Tr(const char* s) {
	QString _ret = QFontComboBox::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QFontComboBox_SetWritingSystem(QFontComboBox* self, int writingSystem) {
	self->setWritingSystem(static_cast<QFontDatabase::WritingSystem>(writingSystem));
}

int QFontComboBox_WritingSystem(const QFontComboBox* self) {
	return self->writingSystem();
}

void QFontComboBox_SetFontFilters(QFontComboBox* self, int filters) {
	self->setFontFilters(static_cast<QFlags<QFontComboBox::FontFilter>>(filters));
}

int QFontComboBox_FontFilters(const QFontComboBox* self) {
	return self->fontFilters();
}

QFont* QFontComboBox_CurrentFont(const QFontComboBox* self) {
	return new QFont(self->currentFont());
}

QSize* QFontComboBox_SizeHint(const QFontComboBox* self) {
	return new QSize(self->sizeHint());
}

void QFontComboBox_SetSampleTextForSystem(QFontComboBox* self, int writingSystem, const libqt_string sampleText) {
	self->setSampleTextForSystem(static_cast<QFontDatabase::WritingSystem>(writingSystem), QString::fromUtf8(sampleText.data, sampleText.len));
}

libqt_string QFontComboBox_SampleTextForSystem(const QFontComboBox* self, int writingSystem) {
	QString _ret = self->sampleTextForSystem(static_cast<QFontDatabase::WritingSystem>(writingSystem));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QFontComboBox_SetSampleTextForFont(QFontComboBox* self, const libqt_string fontFamily, const libqt_string sampleText) {
	self->setSampleTextForFont(QString::fromUtf8(fontFamily.data, fontFamily.len), QString::fromUtf8(sampleText.data, sampleText.len));
}

libqt_string QFontComboBox_SampleTextForFont(const QFontComboBox* self, const libqt_string fontFamily) {
	QString _ret = self->sampleTextForFont(QString::fromUtf8(fontFamily.data, fontFamily.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QFontComboBox_SetDisplayFont(QFontComboBox* self, const libqt_string fontFamily, const QFont* font) {
	self->setDisplayFont(QString::fromUtf8(fontFamily.data, fontFamily.len), *font);
}

void QFontComboBox_SetCurrentFont(QFontComboBox* self, const QFont* f) {
	self->setCurrentFont(*f);
}

void QFontComboBox_CurrentFontChanged(QFontComboBox* self, const QFont* f) {
	self->currentFontChanged(*f);
}

void QFontComboBox_Connect_CurrentFontChanged(QFontComboBox* self, intptr_t slot) {
    void (*slotFunc)(QFontComboBox*, const QFont*) = reinterpret_cast<void (*)(QFontComboBox*, const QFont*)>(slot);
    QFontComboBox::connect(self, &QFontComboBox::currentFontChanged, [self, slotFunc](const QFont& f) {
	slotFunc(self, f);
    });
}

libqt_string QFontComboBox_Tr2(const char* s, const char* c) {
	QString _ret = QFontComboBox::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QFontComboBox_Tr3(const char* s, const char* c, int n) {
	QString _ret = QFontComboBox::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
QSize* QFontComboBox_QBaseSizeHint(const QFontComboBox* self) {
	auto* vqfontcombobox = dynamic_cast<const VirtualQFontComboBox*>(self);
	if (vqfontcombobox && vqfontcombobox->isVirtualQFontComboBox) {
vqfontcombobox->setQFontComboBox_SizeHint_IsBase(true);
	return new QSize(vqfontcombobox->sizeHint());
}
}

// Auxiliary method to allow providing re-implementation
void QFontComboBox_OnSizeHint(const QFontComboBox* self, intptr_t slot) {
	auto* vqfontcombobox = dynamic_cast<const VirtualQFontComboBox*>(self);
	if (vqfontcombobox && vqfontcombobox->isVirtualQFontComboBox) {
vqfontcombobox->setQFontComboBox_SizeHint_Callback(reinterpret_cast<VirtualQFontComboBox::QFontComboBox_SizeHint_Callback>(slot));
}
}

// Derived class handler implementation
bool QFontComboBox_Event(QFontComboBox* self, QEvent* e) {
	auto* vqfontcombobox = dynamic_cast<VirtualQFontComboBox*>(self);
	if (vqfontcombobox && vqfontcombobox->isVirtualQFontComboBox) {
	return vqfontcombobox->event(e);
	} else {
	return self->QFontComboBox::event(e);
}
}

// Base class handler implementation
bool QFontComboBox_QBaseEvent(QFontComboBox* self, QEvent* e) {
	auto* vqfontcombobox = dynamic_cast<VirtualQFontComboBox*>(self);
	if (vqfontcombobox && vqfontcombobox->isVirtualQFontComboBox) {
vqfontcombobox->setQFontComboBox_Event_IsBase(true);
	return vqfontcombobox->event(e);
}
}

// Auxiliary method to allow providing re-implementation
void QFontComboBox_OnEvent(QFontComboBox* self, intptr_t slot) {
	auto* vqfontcombobox = dynamic_cast<VirtualQFontComboBox*>(self);
	if (vqfontcombobox && vqfontcombobox->isVirtualQFontComboBox) {
vqfontcombobox->setQFontComboBox_Event_Callback(reinterpret_cast<VirtualQFontComboBox::QFontComboBox_Event_Callback>(slot));
}
}

void QFontComboBox_Delete(QFontComboBox* self) {
    delete self;
}

File: binding/src/libqfontcombobox.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qfontcombobox.html
pub const qfontcombobox = struct {

    /// New constructs a new QFontComboBox object.
    pub fn New(parent: ?*anyopaque) QtC.QFontComboBox {
        return qtc.QFontComboBox_new(@ptrCast(parent));
    }


    /// New2 constructs a new QFontComboBox object.
    pub fn New2() QtC.QFontComboBox {
        return qtc.QFontComboBox_new2();
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFontComboBox_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfontcombobox.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetWritingSystem(self: ?*anyopaque, writingSystem: i32) void {
        qtc.QFontComboBox_SetWritingSystem(@ptrCast(self), @intCast(writingSystem));
    }

    pub fn WritingSystem(self: ?*anyopaque, ) i32 {
        return qtc.QFontComboBox_WritingSystem(@ptrCast(self));
    }

    pub fn SetFontFilters(self: ?*anyopaque, filters: i32) void {
        qtc.QFontComboBox_SetFontFilters(@ptrCast(self), @intCast(filters));
    }

    pub fn FontFilters(self: ?*anyopaque, ) i32 {
        return qtc.QFontComboBox_FontFilters(@ptrCast(self));
    }

    pub fn CurrentFont(self: ?*anyopaque, ) QtC.QFont {
        return qtc.QFontComboBox_CurrentFont(@ptrCast(self));
    }

    pub fn SizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QFontComboBox_SizeHint(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnSizeHint(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QSize) void {
        qtc.QFontComboBox_OnSizeHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QFontComboBox_QBaseSizeHint(@ptrCast(self));
    }

    pub fn SetSampleTextForSystem(self: ?*anyopaque, writingSystem: i32, sampleText: []const u8) void {
        const sampleText_str = qtc.libqt_string{
    .len = sampleText.len,
    .data = sampleText.ptr,
};
qtc.QFontComboBox_SetSampleTextForSystem(@ptrCast(self), @intCast(writingSystem), sampleText_str);
    }

    pub fn SampleTextForSystem(self: ?*anyopaque, writingSystem: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFontComboBox_SampleTextForSystem(@ptrCast(self), @intCast(writingSystem));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfontcombobox.SampleTextForSystem: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetSampleTextForFont(self: ?*anyopaque, fontFamily: []const u8, sampleText: []const u8) void {
        const fontFamily_str = qtc.libqt_string{
    .len = fontFamily.len,
    .data = fontFamily.ptr,
};
const sampleText_str = qtc.libqt_string{
    .len = sampleText.len,
    .data = sampleText.ptr,
};
qtc.QFontComboBox_SetSampleTextForFont(@ptrCast(self), fontFamily_str, sampleText_str);
    }

    pub fn SampleTextForFont(self: ?*anyopaque, fontFamily: []const u8, allocator: std.mem.Allocator) []const u8 {
        const fontFamily_str = qtc.libqt_string{
    .len = fontFamily.len,
    .data = fontFamily.ptr,
};
const _str = qtc.QFontComboBox_SampleTextForFont(@ptrCast(self), fontFamily_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfontcombobox.SampleTextForFont: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetDisplayFont(self: ?*anyopaque, fontFamily: []const u8, font: ?*anyopaque) void {
        const fontFamily_str = qtc.libqt_string{
    .len = fontFamily.len,
    .data = fontFamily.ptr,
};
qtc.QFontComboBox_SetDisplayFont(@ptrCast(self), fontFamily_str, @ptrCast(font));
    }

    pub fn SetCurrentFont(self: ?*anyopaque, f: ?*anyopaque) void {
        qtc.QFontComboBox_SetCurrentFont(@ptrCast(self), @ptrCast(f));
    }

    pub fn CurrentFontChanged(self: ?*anyopaque, f: ?*anyopaque) void {
        qtc.QFontComboBox_CurrentFontChanged(@ptrCast(self), @ptrCast(f));
    }

    pub fn OnCurrentFontChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QFontComboBox_Connect_CurrentFontChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Event(self: ?*anyopaque, e: ?*anyopaque) bool {
        return qtc.QFontComboBox_Event(@ptrCast(self), @ptrCast(e));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QFontComboBox_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, e: ?*anyopaque) bool {
        return qtc.QFontComboBox_QBaseEvent(@ptrCast(self), @ptrCast(e));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFontComboBox_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfontcombobox.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFontComboBox_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfontcombobox.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QFontComboBox_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/fontcombobox.html#types
pub const enums = struct {
    pub const FontFilter = enum {
        pub const AllFonts: i32 = 0;
        pub const ScalableFonts: i32 = 1;
        pub const NonScalableFonts: i32 = 2;
        pub const MonospacedFonts: i32 = 4;
        pub const ProportionalFonts: i32 = 8;
    };

};
